[#2990] Normalize global CLI args/flags

gshank opened the pull request:
resolves #2990

### Description

Make global CLI arguments/flags consistent by allowing them to be set 1) as a command line argument (before subcommands), 2) as an environment variable  3) in the config part of the profile.

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

gshank commented:
Removed: strict, use_cache, test_new_parser. 

Added command line argument before subcommands:  --no-version-check, --fail-fast, --profiles_dir.

Global CLI flags: use_experimental_parser, debug, log_format, write_json, partial_parse, use_colors, send_anonymous_user_stats, no_version_checks, fail_fast, profiles_dir, printer_width.

All environment variables start with DBT_, like DBT_PROFILES_DIR. 

Note: putting a 'profiles_dir' config in the profile is kind of pointless, but is supported for consistency.

jtcohen6 reviewed (commented):
Really solid start! Excited to clean up our docs for this :)

gshank commented on test/integration/012_deprecation_tests/test_deprecations.py:
```diff
@@ -75,23 +66,15 @@ def project_config(self):
-        with self.assertRaises(dbt.exceptions.CompilationException) as exc:
-            self.run_dbt(strict=True)
-        exc_str = ' '.join(str(exc.exception).split())  # flatten all whitespace
-        assert 'The "adapter_macro" macro has been deprecated' in exc_str
```
Tests have been fixed.

gshank commented on core/dbt/flags.py:
```diff
@@ -50,56 +72,52 @@ def _get_context():
+    lc_flag = flag.lower()
+    flag_value = getattr(args, lc_flag, None)
+    if flag_value is None:
+        env_flag = f"DBT_{flag}"
```
Added a few comments.

gshank commented:
I've merged the flags with the 'args' that show up in run_results.json.  I removed the 'profiles_dir' attribute from UserConfig. Switched to using flags.VERSION_CHECK instead of args. etc etc. Going to review the flags to make sure they're all actually being used and that there are tests, but hopefully functionality is done.

nathaniel-may reviewed (approved):
Oh wow. I didn't realize these were so all over the place. Looks like a solid improvement to me.
